libvast/test/schema.cpp

```diff
@@ -298,6 +296,142 @@ TEST(parseable - complex types global) {
   CHECK(*e == *enum_t);
 }
 
+TEST(parseable - out of order definitions) {
+  using namespace std::string_view_literals;
+  auto str = R"__(
+    type baz = list<bar>
+    type bar = record{
+      x: foo
+    }
+    type foo = int
+  )__"sv;
+  schema sch;
+  CHECK(parsers::schema(str, sch));
+  auto baz = unbox(sch.find("baz"));
+  // clang-format off
+  auto ref = type{
+    list_type{
+      record_type{
+        {"x", integer_type{}.name("foo")}
+      }.name("bar")
+    }.name("baz")
+  };
+  // clang-format on
+  CHECK_EQUAL(baz, ref);
+}
+
+TEST(parseable - with context) {
+  using namespace std::string_view_literals;
+  MESSAGE("prepare the context");
+  auto global = shared_schema_parser::symbol_buffer{};
+  // schema gs;
+  {
+    auto p = shared_schema_parser{global, global};
+    CHECK(p("type foo = count", unused));
+  }
+  {
+    MESSAGE("Use defintion from context");
+    auto str = R"__(
+      type bar = record{
+        x: record{
+          y: foo
+        }
+      }
+    )__"sv;
+    auto local = shared_schema_parser::symbol_buffer{};
+    auto p = shared_schema_parser{global, local};
+    schema sch;
+    CHECK(p(str, sch));
+    auto bar = unbox(sch.find("bar"));
+    // clang-format off
+    auto ref = type{
+      record_type{
+        {"x", record_type{{"y", count_type{}.name("foo")}}}
+      }.name("bar")
+    };
+    // clang-format on
+    CHECK_EQUAL(bar, ref);
+  }
+  {
+    MESSAGE("Override defintion from context - before use");
+    auto str = R"__(
+      type foo = int
+      type bar = record{
+        x: record{
+          y: foo
+        }
+      }
+    )__"sv;
+    auto local = shared_schema_parser::symbol_buffer{};
+    auto p = shared_schema_parser{global, local};
+    schema sch;
+    CHECK(p(str, sch));
+    auto bar = unbox(sch.find("bar"));
+    // clang-format off
+    auto ref = type{
+      record_type{
+        {"x", record_type{{"y", integer_type{}.name("foo")}}}
+      }.name("bar")
+    };
+    // clang-format on
+    CHECK_EQUAL(bar, ref);
+  }
+  {
+    MESSAGE("Override defintion from context - after use");
+    auto str = R"__(
+      type bar = record{
+        x: record{
+          y: foo
+        }
+      }
+      type foo = int
+    )__"sv;
+    auto local = shared_schema_parser::symbol_buffer{};
+    auto p = shared_schema_parser{global, local};
+    schema sch;
+    CHECK(p(str, sch));
+    auto bar = unbox(sch.find("bar"));
+    // clang-format off
+    auto ref = type{
+      record_type{
+        {"x", record_type{{"y", integer_type{}.name("foo")}}}
+      }.name("bar")
+    };
```

Comment:
DRY: you could reuse the baseline definition from above here.

Reply:
In my opinion DRY is a lot less important for unit tests. Minimizing indirections and small scopes provide a better value here. Also, the definition above contains a `count_type` instead of an `integer_type`.